pkg/promlib/library.go

```diff
@@ -0,0 +1,154 @@
+package promlib
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
+	sdkhttpclient "github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/log"
+	"github.com/patrickmn/go-cache"
+	apiv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+
+	"github.com/grafana/grafana/pkg/promlib/client"
+	"github.com/grafana/grafana/pkg/promlib/instrumentation"
+	"github.com/grafana/grafana/pkg/promlib/querydata"
+	"github.com/grafana/grafana/pkg/promlib/resource"
+)
+
+type Service struct {
+	im     instancemgmt.InstanceManager
+	logger log.Logger
+}
+
+type instance struct {
+	queryData    *querydata.QueryData
+	resource     *resource.Resource
+	versionCache *cache.Cache
+}
+
+type extendTransportOptions func(ctx context.Context, settings backend.DataSourceInstanceSettings, clientOpts *sdkhttpclient.Options) (*sdkhttpclient.Options, error)
```

Comment:
couple of nits here:

 - Since this type is used in a exported function, it's better to export it
 - It's not `Transport` specific
 - Since you are passing the pointer to the options, any modifications within the funcion will alter the argument so you don't need to return it (and it's more clear of what it does)
 
So my suggestion: `type ExtendOptions func(ctx context.Context, settings backend.DataSourceInstanceSettings, clientOpts *sdkhttpclient.Options) (error)`

Reply:
https://github.com/grafana/grafana/pull/83952/commits/42130823154657c5ea8b2724f25d14f09579484a